### PR TITLE
refactor(sessiond): Add config flag for max shard size

### PR DESCRIPTION
**Pull request by veshkemburu**
Signed-off-by: veshkemburu <veshkemburu@fb.com>

## Summary

Shard size should be changeable dynamically rather than be set to a constant value. Let's add the field to sessiond.yml

## Test Plan

Checked that proper value was loaded in, and when config was omitted. 

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by pshelar — CHANGES_REQUESTED** [review]
Please post perf numbers with different shard size, so rationalize the default shard size.

File: lte/gateway/c/session_manager/ShardTracker.cpp [hunk 1]
```diff
@@ -14,14 +14,23 @@
...
+#include "includes/ServiceConfigLoader.h"
+#define SESSIOND_SERVICE "sessiond"
 #include "ShardTracker.h"
 
 namespace magma {
 
 ShardTracker::ShardTracker() {
   // initialize with at least one empty vector to avoid checking for empty
   imsis_per_shard_.push_back({});
+  // read max shard size config value from sessiond.yml
```
Comment (koolzz) [inline lte/gateway/c/session_manager/ShardTracker.cpp]: I think its eaasier to just parse config in sessiond_main (and pass the value)

File: cwf/gateway/configs/sessiond.yml [hunk 2]
```diff
@@ -64,3 +64,6 @@ poll_stats_interval: 5
 
 # set to true to enable pull model for stats(polling pipelined from sessiond)
 enable_pull_stats: false
+
+# set to the maximum amount of UEs that can be assigned a particular shard id
```
Comment (themarwhal) [inline cwf/gateway/configs/sessiond.yml]:
```suggestion
maximum number of UEs that can be assigned a particular shard id
```
for all the config files

File: lte/gateway/c/session_manager/ShardTracker.cpp [hunk 3]
```diff
@@ -14,14 +14,14 @@
 #include <string>
 #include <vector>
 #include <set>
-
 #include "ShardTracker.h"
 
 namespace magma {
 
-ShardTracker::ShardTracker() {
+ShardTracker::ShardTracker(uint32_t max_shard_size) {
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/ShardTracker.cpp]:
```suggestion
ShardTracker::ShardTracker(uint32_t max_shard_size): max_shard_size_(max_shard_size) {
```

File: lte/gateway/c/session_manager/sessiond_main.cpp [hunk 4]
```diff
@@ -320,7 +320,14 @@ int main(int argc, char* argv[]) {
 
   // Some setup work for the SessionCredit class
   set_consts(config);
-  auto shard_tracker = std::make_shared<magma::ShardTracker>();
+
+  uint32_t max_shard_size;
+  if (!config["max_shard_size"].IsDefined()) {
+    max_shard_size = 100;
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/sessiond_main.cpp]: Let's define the default value in a variable above like others

Comment (themarwhal) [reply lte/gateway/c/session_manager/sessiond_main.cpp]: If we're using the default value, we should log that

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 5]
```diff
@@ -56,7 +56,7 @@ class LocalEnforcerTest : public ::testing::Test {
     aaa_client           = std::make_shared<MockAAAClient>();
     events_reporter      = std::make_shared<MockEventsReporter>();
     auto default_mconfig = get_default_mconfig();
-    auto shard_tracker   = std::make_shared<ShardTracker>();
+    auto shard_tracker   = std::make_shared<ShardTracker>(100);
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp]: maybe define the default value in Consts.h?

**Review by themarwhal — CHANGES_REQUESTED** [review]
looks good, minor nits.

**Review by themarwhal — APPROVED** [review]

**Review by pshelar — APPROVED** [review]
LGTM

**Comment by ekowtaylor** [comment]
@veshkemburu , @pshelar this PR has now gone stale. To maintain a healthy codebase, it will automatically closed unless there are plans to rebase and merge by close of week

**Comment by pshelar** [comment]
@koolzz can we get this in, I dont see any downside of this.

**Comment by koolzz** [comment]
> @koolzz can we get this in, I dont see any downside of this.

sure, tests seem stuck let me restart

**Comment by ekowtaylor** [comment]
@koolzz, what are the next steps? This has gone stale again

**Comment by ekowtaylor** [comment]
@pshelar , @koolzz this PR has definitely gone stale. Wondering if it makes sense to close and start over if the intent results is still required?

**Pull request closed without merging**